std/array.d

```diff
@@ -778,40 +778,62 @@ private auto arrayAllocImpl(bool minimallyInitialized, T, I...)(I sizes) nothrow
 }
 
 // overlap
-/*
-NOTE: Undocumented for now, overlap does not yet work with ctfe.
+
+
+/++
 Returns the overlapping portion, if any, of two arrays. Unlike $(D
-equal), $(D overlap) only compares the pointers in the ranges, not the
-values referred by them. If $(D r1) and $(D r2) have an overlapping
-slice, returns that slice. Otherwise, returns the null slice.
-*/
-auto overlap(T, U)(T[] r1, U[] r2) @trusted pure nothrow
-if (is(typeof(r1.ptr < r2.ptr) == bool))
+equal), $(D overlap) only compares the pointers and lengths in the
+ranges, not the values referred by them. If $(D r1) and $(D r2) have an
+overlapping slice, returns that slice. Otherwise, returns the null
+slice.
++/
+
+CommonType!(T[], U[]) overlap(T, U)(T[] a, U[] b) @trusted
+if (is(typeof(a.ptr < b.ptr) == bool))
 {
-    import std.algorithm.comparison : min, max;
-    auto b = max(r1.ptr, r2.ptr);
-    auto e = min(r1.ptr + r1.length, r2.ptr + r2.length);
-    return b < e ? b[0 .. e - b] : null;
+    import std.algorithm.comparison : min;
+
+    auto end = min(a.ptr + a.length, b.ptr + b.length);
+    // CTFE requires pairing pointer comparisons, which forces a
+    // slightly inefficient implementation.
+    if (a.ptr <= b.ptr && b.ptr < a.ptr + a.length)
+    {
+        return b.ptr[0 .. end - b.ptr];
+    }
+
+    if (b.ptr <= a.ptr && a.ptr < b.ptr + b.length)
+    {
+        return a.ptr[0 .. end - a.ptr];
+    }
+
+    return null;
```

Comment:
I figured there's another inefficiency in the implementation (caused by my request, sorry). Now we compute `end` compulsively even though it may not be used. Combined with the other inefficiency (redundant comparisons), this makes me a tad unhappy.

So a version with minimal work and no duplication is this:

```d
CommonType!(T*, U*) begin = void;
if (a.ptr <= b.ptr)
{
    if (b.ptr >= a.ptr + a.length)
        return null;
    begin = b.ptr;
}
else
{
    if (a.ptr >= b.ptr + b.length)
        return null;
    begin = a.ptr;
}
return begin[0 .. min(a.ptr + a.length, b.ptr + b.length) - begin];
```

This does three comparisons per call, as it should. According to https://godbolt.org/g/1H23TG, the redundant additions are not optimized away properly resulting in longer code than necessary. This leads to:

```d
    CommonType!(T*, U*) begin = void, endA = void, endB = void;
    if (a.ptr <= b.ptr)
    {
        endA = a.ptr + a.length;
        if (b.ptr >= endA)
            return null;
        begin = b.ptr;
    }
    else
    {
        endB = b.ptr + b.length;
        if (a.ptr >= endB)
            return null;
        begin = a.ptr;
    }
    import std.algorithm;
    return begin[0 .. min(endA, endB) - begin];
```

I'll let this go though.